yt_dlp/extractor/tumblr.py

```diff
@@ -107,116 +237,176 @@ class TumblrIE(InfoExtractor):
-        })
-
-        response, urlh = self._download_webpage_handle(
-            self._LOGIN_URL, None, 'Logging in',
-            data=urlencode_postdata(login_form), headers={
-                'Content-Type': 'application/x-www-form-urlencoded',
-                'Referer': self._LOGIN_URL,
-            })
-
-        # Successful login
-        if '/dashboard' in urlh.geturl():
+        if not self._ACCESS_TOKEN:
+            self.report_warning('Cannot log in without access token')
             return
 
-        login_errors = self._parse_json(
-            self._search_regex(
-                r'RegistrationForm\.errors\s*=\s*(\[.+?\])\s*;', response,
-                'login errors', default='[]'),
-            None, fatal=False)
-        if login_errors:
-            raise ExtractorError(
-                'Unable to login: %s' % login_errors[0], expected=True)
-
-        self.report_warning('Login has probably failed')
+        try:
+            self._download_json(
+                self._OAUTH_URL, None, 'Logging in',
+                data=urlencode_postdata({
+                    'password': password,
+                    'grant_type': 'password',
+                    'username': username,
+                }), headers={
+                    'Content-Type': 'application/x-www-form-urlencoded',
+                    'Authorization': f'Bearer {self._ACCESS_TOKEN}',
+                })
+        except ExtractorError:
+            self.report_warning('Login failed')
 
     def _real_extract(self, url):
         m_url = self._match_valid_url(url)
         video_id = m_url.group('id')
         blog = m_url.group('blog_name')
 
-        url = 'http://%s.tumblr.com/post/%s/' % (blog, video_id)
+        url = f'http://{blog}.tumblr.com/post/{video_id}/'
         webpage, urlh = self._download_webpage_handle(url, video_id)
 
         redirect_url = urlh.geturl()
-        if 'tumblr.com/safe-mode' in redirect_url or redirect_url.startswith('/safe-mode'):
-            raise ExtractorError(
-                'This Tumblr may contain sensitive media. '
-                'Disable safe mode in your account settings '
-                'at https://www.tumblr.com/settings/account#safe_mode',
-                expected=True)
 
+        api_only = bool(self._search_regex(
+            r'(tumblr.com|^)/(safe-mode|login_required|blog/view)',
+            redirect_url, 'redirect', default=None))
+
+        if api_only and not self._ACCESS_TOKEN:
+            raise ExtractorError('Cannot get data for dashboard-only post without access token')
+
+        post_json = traverse_obj(
+            self._download_json(
+                f'https://www.tumblr.com/api/v2/blog/{blog}/posts/{video_id}/permalink',
+                video_id,
+                headers={
+                    'Authorization': f'Bearer {self._ACCESS_TOKEN}',
+                },
+                fatal=False) if self._ACCESS_TOKEN else None,
+            ('response', 'timeline', 'elements', 0), default={})
```

Comment:
sus formatting bro.
Weird that flake8 didn't find it problematic.
But anyways, this is confusing to read.